pkg/discovery/detectors/detector_config.go

```diff
@@ -0,0 +1,101 @@
+package detectors
+
+import (
+	"github.com/golang/glog"
+	"regexp"
+	"strings"
+)
+
+// Master node detection
+// list of node name regexps
+var masterNodeNamePattern *regexp.Regexp
+var masterLabelKeys []*regexp.Regexp
+var masterLabelValues []*regexp.Regexp
+
+// Daemon detection
+// list of node name regexps
+// list of namespace regexps
+var daemonPodNamePattern *regexp.Regexp
+var daemonNamespacePattern *regexp.Regexp
+
+type NodeLabelEntry struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+type MasterNodeDetectors struct {
+	NodeNamePatterns []string         `json:"nodeNamePatterns,omitempty"`
+	NodeLabels       []NodeLabelEntry `json:"nodeLabels,omitempty"`
+}
+
+type DaemonPodDetectors struct {
+	Namespaces      []string `json:"namespaces,omitempty"`
+	PodNamePatterns []string `json:"podNamePatterns,omitempty"`
+}
+
+func ValidateAndParseDetectors(mconfig *MasterNodeDetectors, dconfig *DaemonPodDetectors) error {
+	// Handle default values when sections are missing
+	if mconfig == nil {
+		mconfig = &MasterNodeDetectors{}
+	}
+	if dconfig == nil {
+		dconfig = &DaemonPodDetectors{}
+	}
+
+	// Pre-compile all regular expressions and ensure that they are valid
+
+	// Master node detection by node name
+	masterNodeNamePattern = buildRegexFromList(mconfig.NodeNamePatterns)
+
+	// Master node detection by label
+	masterLabelKeys = make([]*regexp.Regexp, len(mconfig.NodeLabels))
+	masterLabelValues = make([]*regexp.Regexp, len(mconfig.NodeLabels))
+	for i, entry := range mconfig.NodeLabels {
+		masterLabelKeys[i] = regexp.MustCompile(entry.Key)
```

Comment:
It seems that ``regex.MustCompile`` will panic if the expression cannot be parsed? Is it the intention to crash kubeturbo when there is invalid configuration? Or shall we use ``regex.Compile`` to catch and log the error and exit more elegantly?

Reply:
Yes, it's my intention to crash.  I normally wouldn't do this, but I wanted a faulty configuration be be obvious.  If I simply generated an error in the log, nobody would ever notice the problem.  If the pod repeatedly crashes, it will cause the user to review the logs.

Reply:
I just reread your comment, and yes, I agree that I should at least generate a log message before crashing/exiting.

Reply:
In my opinion, if the regex is not parsed, we should use default value instead of crashing it. The fact that the regex doesn't work is not a catastrophe.

Reply:
If the regex doesn’t parse there could be undesirable consequences. User might have made a typo in the master node pattern, if kubeturbo does not exit, the master node will be controllable.

Reply:
I would prefer to crash in this case. This is some error in the configMap. If kubeturbo crashes fast, we can notice it immediately and correct the error in the config.

Reply:
As I stated earlier, we want the error to be obvious.  In addition, the various hosted Kubernetes clusters classify daemons and masters in different ways, so there are no reasonable defaults.

I now terminate kubeturbo by logging an error then using os.Exit(), so I am marking this resolved.